File: backend/src/repositories/delivery.repository.js
const pool = require('../config/database');

// Règle métier : max 10 livraisons actives par livreur
const MAX_DELIVERIES_PER_LIVREUR = 10;

class DeliveryRepository {
  async findById(id) {
    const result = await pool.query(
      `SELECT d.*,
              o.status as order_status,
              u.name as livreur_name, u.email as livreur_email
       FROM deliveries d
       JOIN orders o ON d.order_id = o.id
       LEFT JOIN users u ON d.livreur_id = u.id
       WHERE d.id = $1`,
      [id]
    );
    return result.rows[0] || null;
  }

  async findByOrderId(orderId) {
    const result = await pool.query(
      `SELECT d.*, u.name as livreur_name
       FROM deliveries d
       LEFT JOIN users u ON d.livreur_id = u.id
       WHERE d.order_id = $1`,
      [orderId]
    );
    return result.rows[0] || null;
  }

  async findByLivreur(livreurId) {
    const result = await pool.query(
      `SELECT d.*, o.total_amount
       FROM deliveries d
       JOIN orders o ON d.order_id = o.id
       WHERE d.livreur_id = $1
       ORDER BY d.created_at DESC`,
      [livreurId]
    );
    return result.rows;
  }

  async countActiveByLivreur(livreurId) {
    const result = await pool.query(
      `SELECT COUNT(*) FROM deliveries
       WHERE livreur_id = $1 AND status NOT IN ('livree', 'annulee')`,
      [livreurId]
    );
    return parseInt(result.rows[0].count, 10);
  }

  async isLivreurAvailable(livreurId) {
    const count = await this.countActiveByLivreur(livreurId);
    return count < MAX_DELIVERIES_PER_LIVREUR;
  }

  async create({ orderId, livreurId, address }) {
    const result = await pool.query(
      `INSERT INTO deliveries (order_id, livreur_id, address, status)
       VALUES ($1, $2, $3, 'en_attente')
       RETURNING *`,
      [orderId, livreurId, address]
    );
    return result.rows[0];
  }

  async updateStatus(id, status, { position } = {}) {
    const result = await pool.query(
      `UPDATE deliveries
       SET status = $1,
           current_position = COALESCE($2, current_position),
           delivered_at = CASE WHEN $1 = 'livree' THEN NOW() ELSE delivered_at END,
           updated_at = NOW()
       WHERE id = $3
       RETURNING *`,
      [status, position || null, id]
    );
    return result.rows[0] || null;
  }

  async findAvailableLivreurs() {
    const result = await pool.query(
      `SELECT u.id, u.name, u.email,
              COUNT(d.id) FILTER (WHERE d.status NOT IN ('livree', 'annulee')) as active_deliveries
       FROM users u
       LEFT JOIN deliveries d ON d.livreur_id = u.id
       WHERE u.role = 'livreur'
       GROUP BY u.id, u.name, u.email
       HAVING COUNT(d.id) FILTER (WHERE d.status NOT IN ('livree', 'annulee')) < $1
       ORDER BY active_deliveries ASC`,
      [MAX_DELIVERIES_PER_LIVREUR]
    );
    return result.rows;
  }
}

module.exports = new DeliveryRepository();

File: backend/repositories/delivery_repository.py
from typing import Optional
from sqlalchemy.orm import Session
from sqlalchemy import func
from models.delivery import Delivery, Order
from models.user import User

# Règle métier : max 10 livraisons actives par livreur
MAX_DELIVERIES_PER_LIVREUR = 10
INACTIVE_STATUSES = ("livree", "annulee")


class DeliveryRepository:

    def find_by_id(self, db: Session, delivery_id: str) -> Optional[dict]:
        row = (
            db.query(Delivery, Order.status.label("order_status"),
                     User.name.label("livreur_name"), User.email.label("livreur_email"))
            .join(Order, Delivery.order_id == Order.id)
            .outerjoin(User, Delivery.livreur_id == User.id)
            .filter(Delivery.id == delivery_id)
            .first()
        )
        if not row:
            return None
        d, order_status, livreur_name, livreur_email = row
        data = {c.name: getattr(d, c.name) for c in d.__table__.columns}
        data["order_status"] = order_status
        data["livreur_name"] = livreur_name
        data["livreur_email"] = livreur_email
        return data

    def find_by_order_id(self, db: Session, order_id: str) -> Optional[Delivery]:
        return db.query(Delivery).filter(Delivery.order_id == order_id).first()

    def find_by_livreur(self, db: Session, livreur_id: str):
        rows = (
            db.query(Delivery, Order.total_amount.label("total_amount"))
            .join(Order, Delivery.order_id == Order.id)
            .filter(Delivery.livreur_id == livreur_id)
            .order_by(Delivery.created_at.desc())
            .all()
        )
        result = []
        for d, total_amount in rows:
            data = {c.name: getattr(d, c.name) for c in d.__table__.columns}
            data["total_amount"] = total_amount
            result.append(data)
        return result

    def count_active_by_livreur(self, db: Session, livreur_id: str) -> int:
        return (
            db.query(func.count(Delivery.id))
            .filter(
                Delivery.livreur_id == livreur_id,
                Delivery.status.notin_(INACTIVE_STATUSES),
            )
            .scalar()
        )

    def is_livreur_available(self, db: Session, livreur_id: str) -> bool:
        return self.count_active_by_livreur(db, livreur_id) < MAX_DELIVERIES_PER_LIVREUR

    def create(self, db: Session, order_id: str, livreur_id: str, address: str) -> Delivery:
        delivery = Delivery(
            order_id=order_id,
            livreur_id=livreur_id,
            address=address,
            status="assignee",
        )
        db.add(delivery)
        db.commit()
        db.refresh(delivery)
        return delivery

    def update_status(self, db: Session, delivery_id: str, status: str,
                      position: Optional[str] = None) -> Optional[Delivery]:
        delivery = db.query(Delivery).filter(Delivery.id == delivery_id).first()
        if not delivery:
            return None
        delivery.status = status
        if position:
            delivery.current_position = position
        if status == "livree":
            from sqlalchemy.sql import func as sqlfunc
            delivery.delivered_at = sqlfunc.now()
        db.commit()
        db.refresh(delivery)
        return delivery

    def find_available_livreurs(self, db: Session):
        subq = (
            db.query(
                Delivery.livreur_id,
                func.count(Delivery.id).label("active_count"),
            )
            .filter(Delivery.status.notin_(INACTIVE_STATUSES))
            .group_by(Delivery.livreur_id)
            .subquery()
        )
        rows = (
            db.query(User, func.coalesce(subq.c.active_count, 0).label("active_deliveries"))
            .outerjoin(subq, User.id == subq.c.livreur_id)
            .filter(User.role == "livreur")
            .filter(func.coalesce(subq.c.active_count, 0) < MAX_DELIVERIES_PER_LIVREUR)
            .order_by("active_deliveries")
            .all()
        )
        return [
            {"id": u.id, "name": u.name, "email": u.email, "active_deliveries": int(count)}
            for u, count in rows
        ]


delivery_repo = DeliveryRepository()
